### ISSUE TITLE
invalid regular expression: invalid escape \ sequence

### ISSUE THREAD

**Issue by sephii** [WeblateOrg/weblate#1666]
### Steps to reproduce
I'm not sure how to reproduce, I think this happens when you have accented characters in the dictionary and you try to go to this URL: https://weblate.example.com/translate/myproject/myproject/de/?type=nottranslated. The [dictionary query](https://github.com/WeblateOrg/weblate/blob/weblate-2.17/weblate/trans/models/dictionary.py#L180) that gets built uses `re.escape` to escape the words, and as stated in the [Python docs](https://docs.python.org/3/library/re.html#re.escape):

> Escape all the characters in pattern except ASCII letters, numbers and '_'.

This means characters such as accented characters will be escaped as well and will cause the query to fail (with the error message from PostgreSQL "invalid regular expression: invalid escape \ sequence"). Looking at the [PostgreSQL docs](https://www.postgresql.org/docs/9.3/static/functions-matching.html) this seems indeed illegal:

> A \ followed by an alphanumeric character but not constituting a valid escape is illegal in AREs.

Executing this query produces an error from PostgreSQL. I guess the regular expression escaping shouldn't be done on Python's side but rather on the database side, since each database engine could have different escape rules.

The generated query looks like this:

```
SELECT "trans_dictionary"."id", "trans_dictionary"."project_id", "trans_dictionary"."language_id", "trans_dictionary"."source", "trans_dictionary"."target" FROM "trans_dictionary" WHERE ("trans_dictionary"."project_id" = 2 AND "trans_dictionary"."language_id" = 46 AND "trans_dictionary"."source"::text ~* '^(word\ 1|word\ with\ an\ accent\ \é)$') ORDER BY "trans_dictionary"."source" ASC;
```

[The following thread](https://www.postgresql.org/message-id/4F47C2D6.8000207%40gmail.com) on the PostgreSQL mailing list seems to refer to the same problem.

### Actual behaviour
The page fails with a 500 error.

### Expected behaviour
The page should display the current translations.

### Server configuration

 * Weblate weblate-2.17-5-g394830b0c
 * Python 2.7.13
 * Django 1.10.7
 * six 1.10.0
 * social-auth-core 1.3.0
 * social-auth-app-django 1.2.0
 * django-appconf 1.0.1
 * Translate Toolkit 2.1.0
 * Whoosh 2.7.0
 * defusedxml 0.4.1
 * Git 2.11.0
 * Pillow (PIL) 1.1.7
 * dateutil 2.5.3
 * lxml 3.7.1
 * django-crispy-forms 1.6.1
 * compressor 2.1.1
 * djangorestframework 3.4.0
 * pytz 2016.7
 * pyuca N/A
 * pyLibravatar N/A
 * PyYAML 3.12
 * Mercurial 4.0
 * Database backends: django.db.backends.postgresql_psycopg2

**Comment by nijel** [comment]
You're right with the escaping, it should reflect database needs, while re.escape does escape for Python re engine.

The regexp search was added in 2.17:

https://github.com/WeblateOrg/weblate/blob/9626b914889fb92c4911605f9db694267ab94cb0/weblate/trans/models/dictionary.py#L178-L182

As it operates much faster than previously used one using multiple conditions:

https://github.com/WeblateOrg/weblate/blob/b4eed63fb000a571282e81294d1887474ea238e4/weblate/trans/models/dictionary.py#L178-L181

Apparently it causes problems with postgresql in some cases.

PS: There is SO question on this topic, unfortunately without answer so far: https://stackoverflow.com/q/46083297/225718

**Comment by elarzee** [comment]
this bug affected dictionaries in all languages for a large project, so i came up with this
**desperate stunt  workaround**:
<pre>dictionary = dictionary.filter(
    source__iregex=r'^({0})$'.format(
        '|'.join([word for word in words])
    )
)</pre>
works for us, without side-effects so far.
nonetheless i would very much appreciate a solid solution. please?

**Comment by nijel** [comment]
Indeed this works as long as the string doesn't contain something what is interpreted in the regexp.

**Comment by nijel** [comment]
Thank you for your report, the issue you have reported has just been fixed. 

* In case you see problem with the fix, please comment on this issue. 
* In case you see similar problem, please open separate issue.
* If you are happy with the outcome, consider supporting Weblate by [donating](https://weblate.org/donate/).

**Comment by elarzee** [comment]
as you said, a regex-relevant character in the source string breaks the thing again (we use leading asterisks to mark untranslated strings in the frontend).
now, i do a <code>word.replace('*', '\\\*')</code>, and it works again, but this is getting ugly...

**Comment by nijel** [comment]
This has been already fixed in git, you can apply 7425a749b44abafe36d8f1c9db018f57684e5983

**Comment by elarzee** [comment]
a.. sry, of course.
thanks!
